drv/hisi_sec.c

```diff
@@ -756,3 +759,304 @@ static struct wd_digest_driver hisi_digest_driver = {
+		if (msg->assoc_bytes > 0)
+			flags |= 0x01 << IV_FLAGS_OFFSET;
+
+		msg->aiv[0] = flags;
+		/*
+		  * the last 32bit is counter's initial number,
+		  * but the nonce uses the first 16bit
+		  * the tail 16bit fill with the cipher length
+		  */
+		msg->aiv[msg->iv_bytes - IV_LAST_BYTE2] =
+			msg->in_bytes & IV_LAST_BYTE2_MASK;
+		msg->aiv[msg->iv_bytes - IV_LAST_BYTE1] =
+			msg->in_bytes & IV_LAST_BYTE1_MASK;
+	}
+}
+
+static void fill_aead_bd2_addr(struct wd_aead_msg *msg,
+		struct hisi_sec_sqe *sqe)
+{
+	__u64 addr;
+
+	sqe->type2.data_src_addr = (__u64)msg->in;
+	sqe->type2.data_dst_addr = (__u64)msg->out;
+
+	/* AEAD input MAC addr use in addr */
+	if (msg->op_type == WD_CIPHER_DECRYPTION_DIGEST) {
+	    addr = (__u64)msg->in + msg->in_bytes + msg->assoc_bytes;
+		sqe->type2.mac_addr = addr;
+	}
+
+	/* AEAD output MAC addr use out addr */
+	if (msg->op_type == WD_CIPHER_ENCRYPTION_DIGEST) {
+	    addr = (__u64)msg->out + msg->out_bytes - msg->auth_bytes;
+		sqe->type2.mac_addr = addr;
+	}
+
+	sqe->type2.c_key_addr = (__u64)msg->ckey;
+	sqe->type2.a_key_addr = (__u64)msg->akey;
+	sqe->type2.c_ivin_addr = (__u64)msg->iv;
+
+	/* CCM/GCM should init a_iv */
+	set_aead_auth_iv(msg);
+
+	sqe->type2.a_ivin_addr = (__u64)msg->aiv;
+}
+
+int hisi_sec_aead_send(handle_t ctx, struct wd_aead_msg *msg)
+{
+	handle_t h_qp = (handle_t)wd_ctx_get_priv(ctx);
+	struct hisi_sec_sqe sqe;
+	__u8 scene, cipher, de, auth;
+	__u16 count = 0;
+	int ret;
+
+	if (!msg) {
+		WD_ERR("failed to check input aead msg!\n");
+		return -EINVAL;
+	}
+
+	memset(&sqe, 0, sizeof(struct hisi_sec_sqe));
+	/* config BD type */
+	sqe.type_auth_cipher = BD_TYPE2;
+	/* config scence */
+	scene = SEC_IPSEC_SCENE << SEC_SCENE_OFFSET;
+	de = DATA_DST_ADDR_ENABLE << SEC_DE_OFFSET;
+	auth = AUTH_HMAC_CALCULATE << SEC_AUTH_OFFSET;
+	sqe.type_auth_cipher |= auth;
+
+	if (msg->op_type == WD_CIPHER_ENCRYPTION_DIGEST) {
+		cipher = SEC_CIPHER_ENC << SEC_CIPHER_OFFSET;
+		sqe.sds_sa_type = WD_CIPHER_THEN_DIGEST;
+	} else if (msg->op_type == WD_CIPHER_DECRYPTION_DIGEST) {
+		cipher = SEC_CIPHER_DEC << SEC_CIPHER_OFFSET;
+		sqe.sds_sa_type = WD_DIGEST_THEN_CIPHER;
+	} else {
+		WD_ERR("failed to check aead op type!\n");
+		return -EINVAL;
+	}
+	sqe.sds_sa_type |= (__u8)(de | scene);
+	sqe.type_auth_cipher |= cipher;
+
+	if (msg->in_bytes == 0 ||
+		msg->in_bytes > MAX_INPUT_DATA_LEN) {
+		WD_ERR("failed to check aead input data length!\n");
+		return -EINVAL;
+	}
+	sqe.type2.clen_ivhlen = msg->in_bytes;
+	sqe.type2.cipher_src_offset = msg->assoc_bytes;
+	sqe.type2.alen_ivllen = msg->in_bytes + msg->assoc_bytes;
+
+	ret = fill_aead_bd2_alg(msg, &sqe);
+	if (ret) {
+		WD_ERR("failed to fill aead bd alg!\n");
+		return ret;
+	}
+
+	ret = fill_aead_bd2_mode(msg, &sqe);
+	if (ret) {
+		WD_ERR("failed to fill aead bd mode!\n");
+		return ret;
+	}
+
+	fill_aead_bd2_addr(msg, &sqe);
+
+#ifdef DEBUG
+	WD_ERR("Dump aead send sqe-->!\n");
+	sec_dump_bd((unsigned char *)&sqe, SQE_BYTES_NUMS);
+#endif
```

Comment:
please remove these debug code

Reply:
this debug code is necessary for cipher, digest and aead